app/src/main/java/com/eveningoutpost/dexdrip/watch/miband/MiBandService.java

```diff
@@ -253,114 +347,105 @@ public int onStartCommand(Intent intent, int flags, int startId) {
                 break;
             case "update_bg_as_notification":
                 ((MiBandState) mState).setSendReadingSequence();
-                changeState(INIT);
                 break;
+            default:
+                return;
         }
-    }
-
-    private static final boolean isBetweenValidTime(Date startTime, Date endTime, Date currentTime) {
-        //Start Time
-        Calendar StartTime = Calendar.getInstance();
-        StartTime.setTime(startTime);
-        StartTime.set(1, 1, 1);
 
-        Calendar EndTime = Calendar.getInstance();
-        EndTime.setTime(endTime);
-        EndTime.set(1, 1, 1);
-
-        //Current Time
-        Calendar CurrentTime = Calendar.getInstance();
-        CurrentTime.setTime(currentTime);
-        CurrentTime.set(1, 1, 1);
-        if (EndTime.compareTo(StartTime) > 0) {
-            if ((CurrentTime.compareTo(StartTime) >= 0) && (CurrentTime.compareTo(EndTime) <= 0)) {
-                return true;
-            } else {
-                return false;
-            }
-        } else if (EndTime.compareTo(StartTime) < 0) {
-            if ((CurrentTime.compareTo(EndTime) >= 0) && (CurrentTime.compareTo(StartTime) <= 0)) {
-                return false;
-            } else {
-                return true;
-            }
-        } else {
-            return false;
-        }
+        changeState(INIT);
     }
 
-    private static final int compareTime(Date date1, Date date2) {
-        //Start Time
-        Calendar StartTime = Calendar.getInstance();
-        StartTime.setTime(date1);
-        StartTime.set(1, 1, 1);
-
-        Calendar EndTime = Calendar.getInstance();
-        EndTime.setTime(date2);
-        EndTime.set(1, 1, 1);
-
-        return EndTime.compareTo(StartTime);
+    private boolean isStaleReading() {
+        BgReading last = BgReading.last();
+        if (last == null || last.isStale()) {
+            return true;
+        }
+        return false;
     }
 
-
     private long whenToRetryNextBgTimer() {
         final long bg_time;
+
         Calendar expireDate = Calendar.getInstance();
-        expireDate.setTimeInMillis(System.currentTimeMillis() + BG_UPDATE_INTERVAL);
+        long currTimeMillis = JoH.tsl();
+        expireDate.setTimeInMillis(System.currentTimeMillis() + BG_UPDATE_NO_DATA_INTERVAL);
         isNightMode = false;
         if (MiBandEntry.isNightModeEnabled()) {
             int nightModeInterval = MiBandEntry.getNightModeInterval();
             if (nightModeInterval != MiBandEntry.NIGHT_MODE_INTERVAL_STEP) {
-                Date curr = Calendar.getInstance().getTime();
+                Calendar currCal = Calendar.getInstance();
+                Date curr = currCal.getTime();
                 Date start = MiBandEntry.getNightModeStart();
                 Date end = MiBandEntry.getNightModeEnd();
-                Boolean result = isBetweenValidTime(start, end, curr);
+                boolean result = isBetweenValidTime(start, end, curr);
```

Comment:
You don't need `currCal` and `curr` anymore. Hence, simply use `Calender.getInstance().getTime()` here.